### ISSUE TITLE
Persistence doesn't work in Blazing Trains due to use of task.getawaiter, the unreliable WindowUnload event and the lifecycle of service creation.

### ISSUE THREAD

**Issue by lachlanwgordon** [davidwengier/Trains.NET#180]
When playing Blazing trains I want to be able to save the train track so that if I close the browser, when I visit the game again my previous track is still there.

This already works in the WPF version, and it looks like 99% of the code is there for it to run in Blazor but it doesn't work. 

I'm happy to send a PR, but I'm not sure if my approach to async refactoring is to your liking so I'd like your opinion first.

## Asynchronous Loading
In Blazor Game storage the async method `LocalStorageService.GetItemAsync` is called using `.GetAwaiter().GetResult()`
which fails to run because the WASM run time does not support .GetAwaiter() because it is single threaded. 

The best solution for this is to refactor things so that the calls to LocalStorageService happen in async methods, but the call is a long way down a call stack which ultimately starts with the `Game` field initializer in ServiceLocator. 

### Proposed Solution
1. Change IGameStorage `IEnumerable<IEntity> ReadEntities();` to `Task<IEnumerable<IEntity>> ReadEntitiesAsync();`
2. Change IGameBoard `void GameBoard.Initialize(...)` to `Task GameBoard.InitializeAsync(...)` 
3. Move the call to `GameBoard.Initialize()` out of the constructor and into a new method `InitializeAsync`
4. Change Index.razor.cs `protected override void OnInitialized()` to `protected override Task OnInitializedAsync()`
5. After ServiceLocating _game, call `_game.InitializeAsync()`
6. Find an appropriate place in WPF and WinForms apps to call  `_game.InitializeAsync()`

This may seem a bit drastic, and it requires changes to other platforms, but even in the windows versions of Trains, it probably wasn't ideal to be doing File I/O on the main thread from field initializers so this should be a win all round, even if it's insignificant.

## Save not running because WindowUnload is limited
In Index.razor.cs you subscribe to the WindowUnload even which in turn calls  `GameBoard.Dispose();` which saves. WindowUnload is not a very useful place to make calls because:
* It is only called with some kinds of navigation, not when you close the browser
* You only get a very short period of time to exectute code, and it can't be asynchronous

The async/sync boarder is a littly fuzzy here. I think the async call that would fail would be when JS calls C#, If not it's probably when you call `_ = ...SetItemAsync()` with a fire and forget in Blazor Storage.

### Proposed Solution
1. Move the Save logic out of GameBoard.Dispose() and into a Save or SaveAsync() method.
2. Call Save() from somewhere in the Blazor Code. This could be triggered by a save button in the UI, a timer or after every time the user modifies the entities.

Optionally we could also refactor all the levels of saving to happen in async methods. I'm not sure if this actually gains us anything useful but as a rule I like to have unawaited anyc calls happening as close to the top of the call stack as possible, but changing this would mean chaning all platforms.

## Local Storage Service is null if called from Game Constructor but works on subsequent call.
I'm not sure exactly what's going on here but I have a theory:
1. App Starts
2. Source generated Static field services are initialized
3. As part of Game being initialized it tries to locate LocalStorageService using ASP.Net Core IOC, but it is null
4. BlazorGameStorage is resolved and gets assigned to AspNetCoreServices. I don't really know what this means but there's a comment in Program.cs
5. Subsequent calls (added by me, triggered in UI) can succesfully locate LocalStorageService

### Proposed Solution
Do nothing. The refactoring to fix the asynchronous loading will fix this because we'll no longer be trying to call the storage service from a constructor. 

Alternatively we could probably restructure the lifecycle of creating services, but I'm not sure how to do that cleanly, I think Blazored.LocalStorageServices likes to be run with the Asp.Net Core IOC and I'm not sure how to get it running with your source generator IOC.

**Comment by davidwengier** [comment]
So the keen eyed follower, and I sincerely hope there aren't any, will have noticed that the Blazor load/save code has been a random series of untested hacks, blindly pushed with the hope that one day they would magically start working. Why I never bothered to try them locally, or even *gasp* debug the code, is a question best left for season 3 of Side Projects and The Burnt Out Developer. But you raise a good point!

I basically got to the same conclusion you did, that it needed async and therefore would be viral, and stopped without trying. I would happily take a PR if you feel like adding an InitializeAsync method, and in fact its something that probably should be controlled by the host, rather than the Game, anyway. Things like max grid size is currently hard coded in PIxelMapper, and thats just silly, and should be parameters of some initialize method (though that can come later.. much later.. years later.. or tomorrow randomly, thus the whims of the developer.. but I digress).

I think I played with saving and it was also strangely slow in Blazor, though that could be because of hacky sync-over-async issues, so its worth a shot, but I wouldn't be surprised if a timer was not a good solution. Either way, exposing a Save method to the host, and not using Dispose, makes sense and is a nicer pattern anyway. The timer could even be in JavaScript, or Save could be called in the mouse event handlers or something.

Also, from memory, the services aren't quite right for easy saving and loading, because of the use of serialization services, so there is probably a better abstraction that could be put in place. Ideally I think a simple API to retrieve the current game state as a string, and load a state string, would be good. Again, all can come later.

So in summary, feel free to PR something, and if I don't like it, I'll probably merge it anyway and just silently clean it up later. I'm a bad maintainer, and I avoid conflict 🙂

**Comment by lachlanwgordon** [comment]
Oh, I just realised that Blazored.LocalStorage offers sychronous APIs, and I've used them before so I should have known that. I'll try them out instead. Damnit, I planned out that gaseous async for nothing.  

My experience with saving to local storage is that it's very fast, although I've never tried to measure it and speed is subjective. Maybe you've got more data or higher standards than I do.

**Comment by davidwengier** [comment]
> Oh, I just realised that Blazored.LocalStorage offers sychronous APIs

How did I miss this?!

**Comment by lachlanwgordon** [comment]
I'm getting a lot of errors I don't understand. They're probably just because I don't know C# 9 and 10 very well, but I'd love it if you can nudge me in the right direction.

e.g.
<img width="436" alt="Screen Shot 2021-11-04 at 6 27 10 pm" src="https://user-images.githubusercontent.com/29908924/140273931-0e4db650-516c-4be6-929d-533b52273b33.png">

and
<img width="918" alt="Screen Shot 2021-11-04 at 6 29 59 pm" src="https://user-images.githubusercontent.com/29908924/140274086-ba51bd84-fd1c-480d-9d51-32c4bee3637e.png">

I can work out ways around them but do you know what's going on with C# 8/9/10ish  to make this code invalid? I thought List implemented IEnumerable so this code would be valid. How can adding this/me resolve a null issue?

**Comment by davidwengier** [comment]
The null thing is just info, it's not an error. The missing "this" is the error, thanks to my overzealous code style setup :)

As for the other issue, returning the list would be fine, you just have to comment out the `yield return`. Having that present means the method is an iterator method, and you can use a normal return in them.

**Comment by lachlanwgordon** [comment]
OH. You like having thises in your repo. To each their own, I'll comply.

But the return issue stll confuses me, even with the foreach loop removed I still get an error. Returning a list should be easy, right?
<img width="912" alt="Screen Shot 2021-11-04 at 11 01 36 pm" src="https://user-images.githubusercontent.com/29908924/140310482-2652bf01-0faf-4d80-9a53-6e11d8957709.png">

**Comment by davidwengier** [comment]
Oh, there is a yield break there too. Change that to "return null" and it'll work

**Comment by lachlanwgordon** [comment]
> Oh, there is a yield break there too. Change that to "return null" and it'll work

Thanks, that fixes it. Seems I still don't really understand yield but I'm getting closer.

Does having a "yield" keyword anywhere in the method change how the rest of it works? A bit like how making a method async changes everything?

**Comment by davidwengier** [comment]
> Does having a "yield" keyword anywhere in the method change how the rest of it works? A bit like how making a method async changes everything?

Yes, exactly right.
